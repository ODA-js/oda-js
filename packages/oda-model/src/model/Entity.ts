import { Map, Record, Set } from 'immutable';

import { IEntity, IEntityInit, IEntityStore, IEntityTransform, IEntityACL} from '../interfaces/IEntity';
import { IField } from '../interfaces/IField';
import { IPackage } from '../interfaces/IPackage';
import { Persistent } from './Persistent';
import { TransformField } from './utils';
import { IPackageContext } from '../contexts/IPackageContext';
import { IEntityContext } from '../contexts/IEntityContext';

// tslint:disable-next-line:variable-name
export const DefaultEntity: IEntityStore = {
  name: null,
  title: null,
  description: null,
  acl: null,
  plural: null,
  singular: null,
  fields: null,
  storage: null,
  indexed: null,
  relations: null,
  required: null,
};

// tslint:disable-next-line:variable-name
export const EntityTransform: IEntityTransform = {
  fields: TransformField(),
};

// tslint:disable-next-line:variable-name
export const EntityStorage = Record(DefaultEntity);

export class Entity extends Persistent<IEntityInit, IEntityStore, IPackageContext> implements IEntity {
  public get modelType(): 'entity' {
    return 'entity';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get acl() {
    return this.store.get('acl', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get singular(): string {
    return this.store.get('singular', null);
  }
  public get plural(): string {
    return this.store.get('plural', null);
  }
  public get fields(): Map<string, IField> {
    return this.store.get('fields', null);
  }
  public get relations(): Set<string> {
    return this.store.get('relations', null);
  }
  public get required(): Set<string> {
    return this.store.get('required', null);
  }
  public get indexed(): Set<string> {
    return this.store.get('indexed', null);
  }
  protected transform(input: Partial<IEntityInit>): IEntityStore {
    const result: IEntityStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'fields') {
            result.fields = EntityTransform.fields.transform(input.fields, this);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IEntityStore> & Readonly<IEntityStore>): IEntityInit {
    const result: IEntityInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'fields') {
            result.fields = EntityTransform.fields.reverse(input.fields);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init?: Partial<IEntityInit>, context?: IPackageContext) {
    super();
    if (!context && init && init.fields) {
      throw new Error('context must be provided');
    }
    if (context) {
      this.attach(context);
    }
    this.store = new EntityStorage(this.transform(init));
  }
}
