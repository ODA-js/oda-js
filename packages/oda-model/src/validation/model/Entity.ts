import { IModel } from '../interfaces/IModel';
import { Record } from 'immutable';
import { Map, Set } from 'immutable';
import { Persistent } from './Persistent';
import { transformMap, transformSet } from './utils';
import {
  IEntityPropsStore,
  IEntity,
  IEntityProps,
  IEntityTransform,
} from '../interfaces/IEntity';
import { IField } from '../interfaces/IField';


// tslint:disable-next-line:variable-name
export const DefaultEntity: IEntityPropsStore = {
  modelType: 'entity',
  name: null,
  title: null,
  description: null,
  acl: null,
  plural: null,
  singular: null,
  fields: Map<string, IField>(),
  storage: null,
};

// tslint:disable-next-line:variable-name
export const EntityTransform: IEntityTransform = {
  fields: transformMap<IField>(),
};

// tslint:disable-next-line:variable-name
export const EntityStorage = Record(DefaultEntity);

export class Model extends Persistent<IEntityProps, IEntityPropsStore> implements IEntity {
  public get modelType(): 'entity' {
    return 'entity';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get acl() {
    return this.store.get('acl', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get singular(): string {
    return this.store.get('singular', null);
  }
  public get plural(): string {
    return this.store.get('plural', null);
  }
  public get fields(): Map<string, IField> {
    return this.store.get('fields', null);
  }
  public get relations(): Set<string> {
    return this.store.get('relations', null);
  }
  public get required(): Set<string> {
    return this.store.get('required', null);
  }
  public get indexed(): Set<string> {
    return this.store.get('indexed', null);
  }

  protected transform(input: IEntityProps): IEntityPropsStore {
    return {
      ...input,
      fields: EntityTransform.fields.transform(input.fields),
    };
  }
  protected reverse(input: IEntityPropsStore): IEntityProps {
    return {
      name: input.name,
      title: input.name,
      description: input.name,
      plural: input.plural,
      singular: input.singular,
      modelType: 'entity',
      acl: input.acl,
      storage: input.storage,
      fields: EntityTransform.fields.reverse(input.fields),
    };
  }
  constructor(init: IEntityProps) {
    super();
    this.store = new EntityStorage(this.transform(init));
    this.init = new (Record<IEntityProps>(init))();
  }
}
