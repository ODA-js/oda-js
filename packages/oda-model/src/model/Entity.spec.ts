import { Entity } from './Entity';

describe('Entity', () => {
  let entity: Entity;

  beforeAll(() => {
    expect(() => new Entity()).not.toThrow();
  });

  beforeEach(() => {
    entity = new Entity();
  });

  it('create empty', () => {
    expect(entity.modelType).toBe('entity');
    expect(entity.name).toBeNull();
    expect(entity.title).toBeNull();
    expect(entity.description).toBeNull();
    expect(entity.fields).toBeNull();
    expect(entity.acl).toBeNull();
    expect(entity.indexed).toBeNull();
    expect(entity.context).toBeNull();
    expect(entity.plural).toBeNull();
    expect(entity.relations).toBeNull();
    expect(entity.required).toBeNull();
    expect(entity.singular).toBeNull();
  });

  it('update strings', () => {
    expect(() => entity.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
    })).not.toThrow();

    expect(entity.toJS()).toMatchObject({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
    });
  });

  it('toJS with dupes 1', () => {
    expect(() => entity.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
      fields: [
        { name: 'one' },
        { name: 'one' },
      ],
    })).not.toThrow();

    expect(entity.fields.size).toBe(1);

    expect(entity.toJS()).toMatchObject({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
      fields: [
        { name: 'one' },
      ],
    });
  });

  it('toJS with HashMap', () => {
    expect(() => entity.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
      fields: {
        one: { type: 'string' },
        two: {},
      },
    },
    )).not.toThrow();

    expect(entity.fields.size).toBe(2);

    expect(entity.toJS()).toMatchObject({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
      fields: [
        { name: 'one' , type: 'string'},
        { name: 'two'},
      ],
    });
  });
});
