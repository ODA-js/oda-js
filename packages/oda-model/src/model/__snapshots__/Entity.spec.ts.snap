// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity create empty 1`] = `
Entity {
  "store": Immutable.Record {
    "name": null,
    "title": null,
    "description": null,
    "acl": null,
    "plural": null,
    "singular": null,
    "fields": null,
    "storage": null,
    "indexed": null,
    "relations": null,
    "required": null,
  },
}
`;

exports[`Entity create empty 2`] = `Object {}`;

exports[`Entity toJS with HashMap 1`] = `
Object {
  "description": "very cool",
  "fields": Array [
    Object {
      "name": "one",
      "type": "string",
    },
    Object {
      "name": "two",
    },
  ],
  "name": "cool",
  "title": "very cool title",
}
`;

exports[`Entity toJS with dupes 1 1`] = `
Object {
  "description": "very cool",
  "fields": Array [
    Object {
      "name": "one",
    },
  ],
  "name": "cool",
  "title": "very cool title",
}
`;

exports[`Entity updates with null or undefined 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "title": "very cool title",
}
`;

exports[`Entity updates with null or undefined 2`] = `
Object {
  "name": "cool",
}
`;
